### ISSUE TITLE
Subtitles language == Audio language

### ISSUE THREAD

**Issue by faminebadger** [mpv-player/mpv#1774]
Hi,

I watch a lot of Japanese Anime, so I have my settings set as follows:
   alang=Japanese,jpn,ja,English,eng,en
   slang=English,eng,en

This is great, most of the time, but it ends up displaying English subtitles on English tv shows when they're available.

What would be really good would be an option to say "don't show subtitles if subtitle language matches audio language".  That would suppress subtitles for English shows, but leave it working perfectly for Japanese.

What do you think?  Should be quite simple, no? :)

Cheers

**Comment by haasn** [comment]
Personally, I've always thought that it may be a bit strange to choose audio language and subtitle preferences independently for precisely this reason.

Maybe it would be better to associate them as pairs, eg. “try non-english audio and english subtitles first, otherwise try english audio + no subtitles”. No idea how that would even look in terms of the configuration, but semantically something like:

```langs = (!eng, eng), (eng, none)```

**Comment by faminebadger** [comment]
Yes, agreed.  My ideal setting would be native audio + english subs whereever native != English, otherwise english audio.

However, I'm not clear there's a good way to identify the "correct" native audio choice for all the various container file formats.

**Comment by haasn** [comment]
On second thought, my logic would mean that a multi-language release of an english movie would end up picking something like spanish localization + english subtitles rather than trying english.

(Perhaps it would be best to only explicitly state japanese, for that reason)

**Comment by wm4** [comment]
I've discussed this with @divVerent a while ago, and the conclusion was that the tagging of subtitle/audio tracks is way too inconsistent that it can't work.

**Comment by faminebadger** [comment]
Why is why I proposed the simpler option in the description (hide subtitles if languages are the same & not unknown).

That can work easily enough, and it'd work around most of the files I have a problem with.

**Comment by shirorin** [comment]
I'd like something like this, I know the LAV splitter does "this" in an _elegant_ way, maybe you could use it as _inspiration_: https://1f0.de/lav-splitter/lav-splitter-stream-selection/

**Comment by Soukyuu** [comment]
It's not that simple though, in some cases there will be English subtitles for someone speaking in a foreign language in a movie. If mpv doesn't display any English subtitles for English audio, those won't get displayed either. 

Same situation for a "signs" track for an English anime dub for example (who watches those, etc)

**Comment by wm4** [comment]
If someone can provide an exact algorithm how the selection should work, I can implement it. (I don't feel like figuring out the details myself, because then of course someone will complain that it should work differently, and then I'd have to change everything again etc.)

**Comment by Soukyuu** [comment]
I'll give it a try:
- user should be able to put audio/subtitle languages in pairs
- first entry is audio language, second the subtitle
- incomplete pairs are treated as audio language only, disabling subs
- `!` is an optional modifier for forced subs (aka sings?)
- example of pairs: `jpn:eng;eng:!eng;eng`
- matching in order of occurrence

-> if there is japanese audio, select it and enable english subs
-> if not, and there is english audio, select it and enable only the forced english sub
-> if there is no forced english sub, select english and disable subtitles

Basically, the _audio_ should be the selection criterion, not the subs. The separators between the pairs etc are up to you. The forced parameter parsing of course depends on the subber actually setting them, but there is nothing you can do about that. I don't have a dual language release to check it, so maybe someone who has could look at how the signs subtitle track is marked (if at all)

**Comment by Argon-** [comment]
There should be somerhing for audio-is-present-but-specified-sub-is-not or else you have to give a pair for every possibility you like.

**Comment by Soukyuu** [comment]
The problem with, say, "parse once to find an exact match -> if not found, match audio" is that in the example above it would lead to mpv choosing Japanese as audio without subs, even though it would make more sense to choose English instead.

You usually don't have that many languages to choose from, so just adding one or two "incomplete pairs" at the end should be fine.

**Comment by wm4** [comment]
And what about handling of default and forced flags?

**Comment by Soukyuu** [comment]
I'd say user language preferences should override default flags - because it's just a default, in case the user doesn't specify anything. 

Is the forced flags typically used to display "signs" subtitle tracks? Personally, I think forced flag should have higher priority. Any other ideas?

**Comment by wm4** [comment]
What if there are two audio tracks with the same language, and one of them has the default flag?

**Comment by Soukyuu** [comment]
As in, normal track and commentary? Can't think of anything else. 
But in that case, the default (non-commentary) track should be preferred.

So,
- if forced flag present -> use that track
- if no audio/subtitle match -> use default
- if audio/subtitle match -> check if multiple tracks
- if multiple tracks match -> use the one with the default flag, or the first one if default flag not present for that language

**Comment by arthurwolf** [comment]
Was this ever solved/implemented? Anyone has found some kind of solution to this, even a hack/odd one? My ideal solution would be for mpv to call an external script I can specify in config, that way I'd have all the freedom I want (script is called with a list of audio/sub options, and answers the id/number of each to select/use. this is trivial to implement, and gives absolute freedom to do anything, including complex rules, using libraries etc. I'd write something in nodejs that uses my preferences super finely, but people could do bash scripts or whatever too. this isn't very user-friendly, but it gives the advanced users at least *some* option to get the functionality they want).

**Comment by CogentRedTester** [comment]
There are already user scripts that do this. Just looking through the wiki pages there is [trackselect](https://github.com/po5/trackselect) and my own [sub-select](https://github.com/CogentRedTester/mpv-sub-select). I believe I have come across others in the past as well, though I don't remember where to find them.

**Comment by arthurwolf** [comment]
That's exactly what I was asking for/about, thank you so much!

**Comment by arthurwolf** [comment]
@CogentRedTester I copied the .lua file to scripts/ and the .json file to script-opts/, and removed my alang/slang options in config. Reading the README I think it's telling me that should be enough, am I missing something? When I run a file nothing happens it uses the defaults. I'm doing something wrong obviously, but I can't figure out what from the Readme...

(btw, when you say «For example eng? could be used instead of eng so that the DVD language code en is also matched.» do you mean «en?" instead of «eng?»)

**Comment by CogentRedTester** [comment]
I have no idea what your default settings are, or what you're expecting to change, I suggest you make a new issue thread in the script repository to avoid cluttering up this page. As for `eng?`, in Lua pattern syntax a '?' means that the previous character is optional.

**Comment by arthurwolf** [comment]
Will do, thanks!

**Comment by Traneptora** [comment]
`subs-with-matching-audio=no`
